drivers: serial: nrfx_uart: Allow TX only mode

nordic-krch opened the pull request:
Extended nrf_uart driver to support TX only.
When RX pin is not provided then RX is not started at all. This
allows to achieve low power with logging/console enabled.

Similar to #22242.
Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uart: Allow TX only mode

Extended nrf_uart driver to support TX only.
When RX pin is not provided then RX is not started at all. This
allows to achieve low power with logging/console enabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -407,9 +414,13 @@ static int uart_nrfx_tx_abort(struct device *dev)
 static int uart_nrfx_rx_enable(struct device *dev, u8_t *buf, size_t len,
 			       s32_t timeout)
 {
+	__ASSERT(nrf_uart_rx_pin_get(uart0_addr) != NRF_UART_PSEL_DISCONNECTED,
```
Why not `__ASSERT(IS_RX_PIN(),`?

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -919,9 +930,11 @@ static int uart_nrfx_init(struct device *dev)
+		nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_RXDRDY);
 
-	nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STARTRX);
+		nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STARTRX);
```
Same should be done with the STARTRX task triggered in `uart_nrfx_set_power_state`, see [line 1044](https://github.com/zephyrproject-rtos/zephyr/pull/22932/files#diff-24c27a01c8713b6b093dfd80f40883d0R1044).

nordic-krch commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -12,6 +12,13 @@
+#else
+#define RX_PIN NRF_UART_PSEL_DISCONNECTED
+#endif
+#define IS_RX_PIN() (RX_PIN != NRF_UART_PSEL_DISCONNECTED)
```
changed to `RX_PIN_USED`

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -407,9 +414,12 @@ static int uart_nrfx_tx_abort(struct device *dev)
 static int uart_nrfx_rx_enable(struct device *dev, u8_t *buf, size_t len,
 			       s32_t timeout)
 {
+	__ASSERT(RX_PIN_USED(), "TX only UART instance");
```
What about changing it to:
```
if (!RX_PIN_USED()) {
	__ASSERT(false, "TX only UART instance");
	return -ENOTSUP;
}
```
(and using similar code also for UARTE)? The rest of the function would be then optimized out when RX pin is not used.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -1024,7 +1040,10 @@ static void uart_nrfx_set_power_state(struct device *dev, u32_t new_state)
-		nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STARTRX);
+		if (RX_PIN_USED()) {
+			nrf_uart_task_trigger(uart0_addr,
+					      NRF_UART_TASK_STARTRX);
```
UARTE driver needs similar change. I guess it would be okay to do this in this PR.

nordic-krch committed:
drivers: serial: nrfx_uarte: Improve TX only mode

Added return -ENOTSUP in uart_rx_enable() if RX pin is disabled.
Fixed power management call where TASKRX was started even though
RX pin was disabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@ioannisg @carlescufi could you merge?
